sklearn/neural_network/_multilayer_perceptron.py

```diff
@@ -977,17 +982,18 @@ def _validate_input(self, X, y, incremental):
         if y.ndim == 2 and y.shape[1] == 1:
             y = column_or_1d(y, warn=True)
 
-        if not incremental:
-            self._label_binarizer = LabelBinarizer()
-            self._label_binarizer.fit(y)
-            self.classes_ = self._label_binarizer.classes_
-        elif self.warm_start:
+        if self.warm_start and (incremental or hasattr(self, "classes_")):
             classes = unique_labels(y)
             if set(classes) != set(self.classes_):
                 raise ValueError("warm_start can only be used where `y` has "
                                  "the same classes as in the previous "
                                  "call to fit. Previously got %s, `y` has %s" %
                                  (self.classes_, classes))
+
+        if not incremental:
+            self._label_binarizer = LabelBinarizer()
+            self._label_binarizer.fit(y)
+            self.classes_ = self._label_binarizer.classes_
```

Comment:
The issue though is that with a warm start (where now `incremental=False`), the `self.classes_` should not be redefined, unless it was not set previously.

Reply:
Indeed, thanks for the detailed check. I pushed a new commit the reorders the checks in the validate function, and also add a matrix of actions taken to make it easier to see why what is done where.

Reply:
I don't know if this is possible, but we should have had a test failing.

Reply:
I'm not sure now (away from kbd) but I think this test was actually dead code in the previous version, that is the reason it couldn't fail in a test. I guess with the current setup it should be possible to trigger a failure.

Reply:
To be more specific than my previous comment, now that I am back at the kbd:

Before, there was no way that `incremental == True` and `warm_start == False`, because `incremental` was passed in as `self.warm_start and hasattr(self, "classes_")`.

The "dead code" I mentioned above is actually not dead code, but can happen when `MLPClassifier::partial_fit` is called with incorrect initial classes, then this is already tested in `test_mlp.py::test_partial_fit_classes_error`.